masschallenge/impact-api
# [AC-5382] - Bug: Image Proxy View is Broken

**Changes Introduced in [AC-5382](https://masschallenge.atlassian.net/browse/AC-5382):**

- V0 API Proxy View was broken, now it is not.
- profile.image field was incorrectly declared as a CharField, instead of a sorl.thumbnail.ImageField
  - an unused
- startup.high_resolution_logo was formatted incorrectly, now it is.
- sorl.thumbnail was trying to run migrations which led to errors, now it is not.

## Testing  setup
### Set up using test4:
- Accelerate URL: https://accelerate-test-4-elb-2-897264252.us-east-1.elb.amazonaws.com/applicant/
- Impact URL: https://api-test-4-elb-827678781.us-east-1.elb.amazonaws.com/

- in accelerate's admin, give your **you+superuser** v0_clients permissions (i.e. add the v0_clients  group via the "Groups" section on your user's admin page)
- in impact-api, create an application for your user
- get this [postman collection](https://www.getpostman.com/collections/1c3cb99cd15a4c57f592).
- set the postman global variables:
  - username
  - password
  - client_id (for the app you created)
  - impact_api_host = (host from above **api-test-4** URL)
  - accelerate_host = (host from above **accelerate-test-4** URL)
- in postman, call `get_auth_token`.

(see **"Test"** instructions, below "Set up locally")

### Set up locally
- `make checkout branch=AC-5382 # both accelerate and impact should have one`
- find out your VM's IP. On Mac: `export HOST=$(ifconfig | grep -Eo 'inet (addr:)?([0-9]*\.){3}[0-9]*' | grep -Eo '([0-9]*\.){3}[0-9]*'  | grep -v '127.0.0.1' | tail -n1)`.
  - On linux replace `ifconfig` with `ip addr show`
- DB:
  ```
  scp mc-prod:/vol/www/tmp/proddb-2018-04-04-11-10.sql.gz .
  make load-db gz_file=./proddb-2018-04-04-11-10.sql.gz
  ```
- Setup shared local impact (AC-5382) and accelerate (development) + S3 storage:
  - in impact:
    - Change the env vars in impact-api **.dev.env**:
  ```
  ACCELERATE_SITE_URL=http://<the value of $HOST>:8181
  IMPACT_API_V0_IMAGE_PASSWORD=<same as accelerate.settings.IMAGE_TOKEN_PASSWORD>
  IMPACT_API_V0_SECURITY_KEY=<same as mc.models.Site.objects.first().security_key>
  AWS_ACCESS_KEY_ID=<get value from opsworks.test1.stack_settings>
  AWS_SECRET_ACCESS_KEY=<get value from opsworks.test1.stack_settings>
  AWS_STORAGE_BUCKET_NAME=accelerate-assets-testing
  ```
    - add to settings.py, under Dev:
    ```
    AWS_LOCATION = 'media'
    AWS_ACCESS_KEY_ID = os.environ.get('AWS_ACCESS_KEY_ID')
    AWS_SECRET_ACCESS_KEY = os.environ.get('AWS_SECRET_ACCESS_KEY')
    AWS_STORAGE_BUCKET_NAME = os.environ.get('AWS_STORAGE_BUCKET_NAME')
    AWS_S3_CUSTOM_DOMAIN = f'{AWS_STORAGE_BUCKET_NAME}.s3.amazonaws.com'
    AWS_S3_OBJECT_PARAMETERS = {
        'CacheControl': 'max-age=86400',
    }

    MEDIA_URL = f'https://{AWS_S3_CUSTOM_DOMAIN}/{AWS_LOCATION}/'
    MEDIA_ROOT = AWS_LOCATION

    DEFAULT_FILE_STORAGE = 'impact.media_storage_backend.MediaStorageBackend'
    STATICFILES_STORAGE = (
        'django.contrib.staticfiles.storage.StaticFilesStorage')

    ```
  - in accelerate:
    - in **develop_settings.py** add to `ALLOWED_HOSTS` the value of `$HOST`.
    - comment out the *settings.py* that are duplicated: `MEDIA_ROOT`, `MEDIA_URL` and `DEFAULT_FILE_STORAGE`.
   - set in *develop_settings.py* the following:
```
AWS_LOCATION = 'media'
AWS_ACCESS_KEY_ID = #AWS_ACCESS_KEY_ID#
AWS_SECRET_ACCESS_KEY = #AWS_SECRET_ACCESS_KEY#
AWS_STORAGE_BUCKET_NAME = 'accelerate-assets-testing'
AWS_S3_CUSTOM_DOMAIN = 's3.amazonaws.com/{AWS_STORAGE_BUCKET_NAME}'.format(
    AWS_STORAGE_BUCKET_NAME=AWS_STORAGE_BUCKET_NAME)
AWS_S3_OBJECT_PARAMETERS = {
    'CacheControl': 'max-age=86400',
}

MEDIA_URL = 'https://{AWS_S3_CUSTOM_DOMAIN}/{AWS_LOCATION}/'.format(
    AWS_S3_CUSTOM_DOMAIN=AWS_S3_CUSTOM_DOMAIN, AWS_LOCATION=AWS_LOCATION)
MEDIA_ROOT = AWS_LOCATION

DEFAULT_FILE_STORAGE = 'mcproject.media_storage_backend.MediaStorageBackend'
STATICFILES_STORAGE = (
    'django.contrib.staticfiles.storage.StaticFilesStorage')

```

- run `make build-all;make run-all-servers`
- in accelerate's admin, give your you+superuser v0_clients permissions.
- in impact-api, create an application for your user
- get this [postman collection](https://www.getpostman.com/collections/1c3cb99cd15a4c57f592).
- set the postman global variables:
  - username
  - password
  - client_id (for the app you created)
  - impact_api_host = http://localhost:8000
  - accelerate_host = http://localhost:8181
- in postman, call `get_auth_token`.

## Test
- run `get startup detail (new)`.
  - on development, you'll see that `logo_url` has incorrect url (not an s3 url). On branch correct.
  - on development, you'll see that team_members' `photo_url` is likewise incorrect. On branch correct.
  - copy the `image_token` value. go to `get image (new)`. paste the value in the ImageToken query parameter, and execute. on development, this fails, on branch it succeeds (returns the aquaGen logo).


## Teardown
- `rm proddb-2018-04-04-11-10.sql.gz`
- `load-db`
- revert the env vars that you set in **.dev.env** and **develop_settings.py** , specifically those that have keys in them.
